src/sdk/pynni/nni/bohb_advisor/bohb_advisor.py

```diff
@@ -0,0 +1,587 @@
+    """
+    A bracket in BOHB, all the information of a bracket is managed by 
+    an instance of this class.
+
+    Parameters
+    ----------
+    s: int
+        The current SH iteration index.
+    s_max: int
+        total number of SH iterations
+    eta: float
+        In each iteration, a complete run of sequential halving is executed. In it,
+		after evaluating each configuration on the same subset size, only a fraction of
+		1/eta of them 'advances' to the next round.
+	max_budget : float
+		The largest budget to consider. Needs to be larger than min_budget!
+		The budgets will be geometrically distributed
+        :math:`a^2 + b^2 = c^2 \sim \eta^k` for :math:`k\in [0, 1, ... , num\_subsets - 1]`.
+    optimize_mode: str
+        optimize mode, 'maximize' or 'minimize'
+    """
+    def __init__(self, s, s_max, eta, max_budget, optimize_mode):
+        self.s = s
+        self.s_max = s_max
+        self.eta = eta
+        self.max_budget = max_budget
+        self.optimize_mode = optimize_mode
+
+        self.n = math.ceil((s_max + 1) * eta**s / (s + 1))
+        self.r = math.ceil(max_budget / eta**s)
+        self.i = 0
+        self.hyper_configs = []         # [ {id: params}, {}, ... ]
+        self.configs_perf = []          # [ {id: [seq, acc]}, {}, ... ]
+        self.num_configs_to_run = []    # [ n, n, n, ... ]
+        self.num_finished_configs = []  # [ n, n, n, ... ]
+        self.no_more_trial = False
+
+    def is_completed(self):
+        """check whether this bracket has sent out all the hyperparameter configurations"""
+        return self.no_more_trial
+
+    def get_n_r(self):
+        """return the values of n and r for the next round"""
+        return math.floor(self.n / self.eta**self.i), self.r * self.eta**self.i
+
+    def increase_i(self):
+        """i means the ith round. Increase i by 1"""
+        self.i += 1
+
+    def set_config_perf(self, i, parameter_id, seq, value):
+        """update trial's latest result with its sequence number, e.g., epoch number or batch number
+
+        Parameters
+        ----------
+        i: int
+            the ith round
+        parameter_id: int
+            the id of the trial/parameter
+        seq: int
+            sequence number, e.g., epoch number or batch number
+        value: int
+            latest result with sequence number seq
+        Returns
+        -------
+        None
+        """
+        if parameter_id in self.configs_perf[i]:
+            if self.configs_perf[i][parameter_id][0] < seq:
+                self.configs_perf[i][parameter_id] = [seq, value]
+        else:
+            self.configs_perf[i][parameter_id] = [seq, value]
+
+    def inform_trial_end(self, i):
+        """If the trial is finished and the corresponding round (i.e., i) has all its trials finished,
+        it will choose the top k trials for the next round (i.e., i+1)
+
+        Parameters
+        ----------
+        i: int
+            the ith round
+        """
```

Comment:
Please specific return value

Reply:
Add already

Reply:
Have you done this? I didn't see this in the commit.

Reply:
??? please check the latest version...The content is:

new trial or None:
            If we have generated new trials after this trial end, we will return a new trial parameters.
            Otherwise, we will return None.

Reply:
Sorry, I see the wrong version. 
Just forget it :-)